app/src/main/java/waiter/Routes.java

```diff
@@ -2,41 +2,38 @@
 
 import java.util.HashMap;
 import java.util.Map;
+import java.util.concurrent.Callable;
 
 public class Routes {
 
-    public Map<String, Map<String, Map<String, String>>> routes;
+    public Map<String, Route> routes;
 
     public Routes() {
         this.routes = new HashMap<>();
     }
 
-    private void addRoute(Route route) {
-        this.routes.putIfAbsent(route.getKey(), route.getValue());
-        if(this.routes.containsKey(route.getKey())) {
-            this.routes.get(route.getKey()).putAll(route.getValue());
-        }
+    public void addRoute(Route route) {
+        this.routes.put(route.getUrl(), route);
     }
 
-    public Routes constructRoutes() {
-        RouteDirector routeDirector = new RouteDirector();
-        RouteBuilder routeBuilder = new RouteBuilder();
-
-        this.addRoute(routeDirector.constructGetSimpleGet(routeBuilder));
-        this.addRoute(routeDirector.constructHeadSimpleGet(routeBuilder));
-        this.addRoute(routeDirector.constructGetSimpleGetWithBody(routeBuilder));
-        this.addRoute(routeDirector.constructNoUrl(routeBuilder));
+    public boolean exists(Request request) {
+        return this.routes.containsKey(request.getUrl());
+    }
 
-        for (String url : this.routes.keySet()) {
-            this.addRoute(routeDirector.constructUrlExistsButNoMethodFound(routeBuilder, url));
+    public Response handle(Request request) {
+        Route route = this.routes.get(request.getUrl());
+        if(!route.methodExistsForUrl(request.getMethod())) {
```

Comment:
Consider moving this to `Router` since it's already handling the route not found scenario. Having both `Router` and `Route` handle requests don't feel right. IMO `Router` should decide how each route is handled, `Routes` is a collection of `Route` and so it knows about all the routes there are and can provide information to the `Router` to help make the decision.

Reply:
Makes sense to me! Thank you 😄

Reply:
Addressed in [this commit](https://github.com/jonmanahan/waiter/pull/18/commits/509da0f689e34e8dd6bf4b266f5885afb275fde5)

Reply:
👍 Btw, feel free to spin up follow-up tickets to address concerns/comments that aren't immediately relevant to this PR. I definitely noticed the number of comments and don't want to make this be the "make the code perfect" PR.